google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionImpl.java

```diff
@@ -574,6 +578,31 @@ Timestamp getCommitTimestampOrNull() {
         : this.currentUnitOfWork.getCommitTimestampOrNull();
   }
 
+  @Override
+  public CommitResponse getCommitResponse() {
+    ConnectionPreconditions.checkState(!isClosed(), CLOSED_ERROR_MSG);
+    ConnectionPreconditions.checkState(
+        this.currentUnitOfWork != null, "There is no transaction on this connection");
+    return this.currentUnitOfWork.getCommitResponse();
+  }
+
+  CommitResponse getCommitResponseOrNull() {
```

Comment:
I still don't see why both getCommitResponseOrNull and getCommitResponse exist. Why not just getCommitResponse that returns null if there's no unit of work?

Reply:
These two methods serve two different APIs that behave differently:
* The `Connection` interface contains several methods that return the current state of the connection or of the last statement. The chosen pattern for that API is to throw an exception if the `Connection` is not in a valid state for the method to be called. I don't think we should deviate from that pattern for one method. This also somewhat reflects the standard in JDBC, where most methods also throw an exception if the `Connection` is not in a valid state for the method. See for example the [commit()](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#commit--) method. (Although this is not 100% true in all cases, as some other JDBC methods have a different contract, and allows you to call it in what seems to be an illogical state. Calling [setAutoCommit](https://docs.oracle.com/javase/8/docs/api/java/sql/Connection.html#setAutoCommit-boolean-) is for example allowed during a transaction, and will automatically commit the transaction __if__ the auto commit mode).
* The Connection API also contains a simple SQL parser that can be used to get and set the state of a connection. This 'API' is lenient, and will return `null` when a value is requested that is not available at that moment. This SQL parser uses the `getCommitResponseOrNull` and other similar methods.